src/concepts/mods.md

```diff
@@ -1 +1,506 @@
+whenever discussion the equality of expressions. Under the eyes of the language,
+equal expressions are just that - equal, and there is no need to disambiguate
+them.
+
+This idea, however, is rather limited in scope. What if we would like to
+deal with a function that depends upon another function? Then we cannot freely
+make the swap between functions, and substitute a function into a context where
+its dependencies do not exist. Indeed, we cannot perform such a substitution at
+all in the case where a function is in any way defined beyond its own function
+body itself.
+
+We don't want to outlaw such practices, however - we simply need to make our
+_abstraction boundaries_ more explicit. When we have a package or otherwise
+standalone bundle of software, we should be able to use all of its components
+that are muutally dependent on each other - not just make our distinctions at
+the function level. We will do so using SML's module system.
+
+## Modules: Basics
+
+SML takes both the ideas of the _specification_ and the _implementation_ and
+provides a way of codifying both within the language itself. Of course, it is
+rather difficult to say "implement the list reversal function in \( O(n) \) asymptotic
+time in \( n \), the length of the list" in code<a
+href="#footnote1">_<sup>[1]</sup>_</a>, so SML will only deal with
+specifications at the _type-level_. That is, within SML itself, a specification
+for a function is simply a type for that function.
+
+The term for a specification in SML is called a _signature_. Our running example
+throughout this section will be of the specification and implementation of a
+package for modular arithmetic. Note that the comments are optional.
+
+```sml
+signature MOD_ARITH =
+sig
+    (* mod_add should be such that mod_add n x y = (x + y) mod n *)
+    val mod_add : int -> int -> int -> int
+
+    (* mod_sub should be such that mod_sub n x y = (x - y) mod n *)
+    val mod_sub : int -> int -> int -> int
+
+    (* mod_times should be such that mod_times n x y = (x * y) mod n *)
+    val mod_times : int -> int -> int -> int
+end
+```
+
+This signature defines three functions, all of which have type `int -> int ->
+int -> int`, which takes in the modulus of the operation, as well as the two
+operands to the operation to be performed. While we would like an implementation
+of this signature to also display the behavior outlined in the comments,
+realistically there is no way to verify that this is true (why?), so the best that we
+can do is ensure that an implementation provides functions with the correct
+type.
+
+Note also that this syntax is slightly different than that you have seen before
+- we use `val` as if we were going to produce a val binding (with an associated
+  type annotation), however we do not actually give it any definition. In truth,
+you _cannot_ define a value in the signature itself, since it's impossible in
+general to check if two values are the same. You can, however, define a `type`
```

Comment:
could possibly elaborate on this point (maybe mention equality types, and say if values were in a signature, then the compiler would have to check that multiple structures have the same value, which is not possible)